Add support and stability level definitions (#9412)
open-telemetry/opentelemetry-collector-contrib: README.md

# OpenTelemetry Collector Contrib

> This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. Some of these components are part of the core distribution, such as the Jaeger and Prometheus ones, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.

Reviewer: ```suggestion
This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. Some of these components are part of the core distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.
```

The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository.

> Each component has its own stability and support levels, as defined in the following sections.

Reviewer 2: Should we have a template table/some kind of 'copy and fill in the blanks' to add on components' READMEs?

Author: I'm hoping one of the next PRs will establish the template. From there, we can determine whether it's easy enough for other components to copy/paste, or whether a template somewhere would be better.

Reviewer 2: 👍 added https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9377#issuecomment-1109510746

> ## Stability levels

Reviewer 3: Should we have a way to expose this on the zpages or with a special command on the otelcol binary?

Author: Good call. I think @esnible added an option to print out all the components of the collector upon error, perhaps a code similar to that would be good to have here. I'll create another issue to track this.

Author: Created #9518 and #9519.

> ## Stability levels

Reviewer 2: Do we want a 'Deprecated' stability level or is that out of scope?

Author: Great call, we need deprecated there.

> ## Stability levels

Reviewer 2: What happens if a component is stable for a telemetry signal but not for others?  For example, how does the current status of the OTLP receiver defined [here](https://github.com/open-telemetry/opentelemetry-collector#status) map to these levels? I think it would not make sense to consider it stable unless logs functionality was under a feature flag.

Do we also want to make a distinction between configuration stability and behavior stability (say, do we want to allow expressing 'this component behavior is stable but the way it is configured may change in breaking ways')?

Author: We definitely need a dimension for the signal type, but I don't think it's a good user-experience to break it even more in config vs. behavior stability. From the user's perspective, they are all part of the same contract, IMO.

Author: I added a few words about it, would you mind taking a look?

Reviewer 2: Looks good and agree that the distinction between configuration and behavior may not be a good user experience. I left a comment below but we can resolve this one!

While we intend to provide high-quality components as part of this repository, we acknowledge that not all of them are ready for prime time. As such, each component should list its current stability level, according to the following definitions:

* **In development**: not all pieces of the component are in place yet and it might not be available as part of any distributions yet. Bugs and performance issues should be reported, but it is likely that the component owners might not give them much attention. Your feedback is still desired, especially when it comes to the user-experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component should not be used in production.
* **Experimental**: the component is ready to be used for limited non-critical workloads and the authors of this component would welcome your feedback. Bugs and performance problems should be reported, but component owners might not work on them right away. The configuration options might change between releases, although component owners should try to minimize breaking changes.

> * **Stable**: the component is ready to be used in production. Bugs and performance problems should be reported and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output are not expected to happen without prior notice, unless under special circumstances.

Reviewer 3: Do we need a beta level as well, or RC?

Author: How would you describe them both? In my view, experimental encompasses both, and component maintainers would be free to move from experimental to stable as they see fit.

Author: I would read the alpha stage as "Experimental" here, while their beta would map to "Stable". I don't think the beta would be a stage in between, given the definitions from this PR.

Author: Perhaps it would be better to rename "Experimental" to "Beta"?

Reviewer 4: Maybe breaking changes to the configuration interface can be used as one of differentiator between Experimental and Beta? Current configuration guarantees for Experimental seem too strict IMO.

E.g. **Beta** is ready for production use but the configuration options might change between releases, although component owners should try to minimize breaking changes.

While **Experimental** configuration options might break often depending on how things evolve.

Author: Alright, I renamed Experimental to Alpha and added a Beta. As a user, I still think I would prefer to have a single "yes/no" on whether I should use it in production, as the message here is now:

- yes, if it's stable
- maybe, if it's beta (your config might break)
- no, if it's alpha

> * **Stable**: the component is ready to be used in production. Bugs and performance problems should be reported and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output are not expected to happen without prior notice, unless under special circumstances.

Reviewer 2: Should we explicitly mention that functionality exposed under feature gates is not covered by these stability guarantees?

Author: I'll add a paragraph about it.

Author: Can you check it out?

Reviewer 2: Current wording looks good, I think we should also link to https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/featuregate/README.md#feature-lifecycle to provide more context

> ## Support levels

Reviewer 3: Is there a level as well here? or just a support "party"?

Author: I think we should start with a stability level and support party. Those supporting a component are the ones defining the stability level, so, they can adjust the expectations as needed.

Reviewer 3: I am asking about the title, showed we call this "Support level" or can we find a different word? My fear is to not confuse the "stability **level**" with "support **level**"

Reviewer: Could probably get away with only "Support"

The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository.

Each component has its own support levels, as defined in the following sections. For each signal that a component supports, there's a stability level, setting the right expectations. It is possible then that a component will be **Stable** for traces but **Experimental** for metrics and **In Development** for logs.

## Stability levels

> While we intend to provide high-quality components as part of this repository, we acknowledge that not all of them are ready for prime time. As such, each component should list its current stability level, according to the following definitions:

Reviewer 2: ```suggestion
While we intend to provide high-quality components as part of this repository, we acknowledge that not all of them are ready for prime time. As such, each component should list its current stability level for each telemetry signal, according to the following definitions:
```

> * **In development**: not all pieces of the component are in place yet and it might not be available as part of any distributions yet. Bugs and performance issues should be reported, but it is likely that the component owners might not give them much attention. Your feedback is still desired, especially when it comes to the user-experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component should not be used in production.

Reviewer 5: Should we define a state which precedes "In development"? E.g. when component is not ready to be used in any way and has only interface available (during the good practices of breaking the large work into chunks). Something like "Do not use", "Bootstrapping" or so?

Author: That's precisely what I had in mind for "In development" -- the component is likely not in a working state, it hasn't been finished, and might not even be tied to the factories yet.

Reviewer: What about components that are in the Contrib repository but not part of the contrib distribution (transform processor is an example I believe).  Should it be marked as "In Developement" or is there something even lower?

Reviewer: "it might not be available as part of any distributions yet"  answer was already in there.

# OpenTelemetry Collector Contrib

> This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. Some of these components are part of the core distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.

Reviewer 5: > Some of these components are part of the core distribution, such as the Jaeger and Prometheus components

Not sure if I read it right but Jaeger and Prometheus were moved from core to contrib quite a while ago

Reviewer: Thats true, but that are still included in the core distribution of the collector: https://github.com/open-telemetry/opentelemetry-collector-releases/blob/main/distributions/otelcol/manifest.yaml

Reviewer 5: So what I'm thinking is that perhaps this would be more clear in what's the relationship between core/contrib repositories/releases

```suggestion
This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. 

The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository. Some of the components in this repository are part of the "core" distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.
```

# OpenTelemetry Collector Contrib

This is a repository for OpenTelemetry Collector components that are not suitable for the  [core repository](https://github.com/open-telemetry/opentelemetry-collector) of the collector. 

The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository. Some of the components in this repository are part of the "core" distribution, such as the Jaeger and Prometheus components, but most of the components here are only available as part of the "contrib" distribution. Users of the OpenTelemetry Collector are also encouraged to build their own custom distributions with the [OpenTelemetry Collector Builder](https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder), using the components they need from the core repository, the contrib repository, and possibly third-party or internal repositories.

> 
> The official distributions, core and contrib, are available as part of the [opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository.

Reviewer 5: ```suggestion

```
After merging my previous suggestion (should have used a multi-line one) this is no longer necessary

## Stability levels

While we intend to provide high-quality components as part of this repository, we acknowledge that not all of them are ready for prime time. As such, each component should list its current stability level for each telemetry signal, according to the following definitions:

* **In development**: not all pieces of the component are in place yet and it might not be available as part of any distributions yet. Bugs and performance issues should be reported, but it is likely that the component owners might not give them much attention. Your feedback is still desired, especially when it comes to the user-experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component should not be used in production.
* **Alpha**: the component is ready to be used for limited non-critical workloads and the authors of this component would welcome your feedback. Bugs and performance problems should be reported, but component owners might not work on them right away. The configuration options might change often without backwards compatibility guarantees.

> * **Beta**: same as Alpha, but the configuration options are deemed stable. While there might be breaking changes between releases, component owners should try to minimize them.

Reviewer 4: Didn't you propose that **Beta** can be used in production while config might break? https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9412#discussion_r860038480

Author: I had only the config changes in mind based on previous comments, but we could say that people can have higher confidence in a beta component to run it with more workloads than they'd run with an alpha component. Let me draft something to that effect.

Author: > A component at this stage is expected to have had exposure to non-critical production workloads already during its **Alpha** phase, making it suitable for broader usage.

How about this?

Reviewer: I think that wording is good.  I like keeping "Stable" as the only place to official say "this is officially GA".

Reviewer: Although now I'm wondering if instead of "**Stable**: the component is ready to be used in production." we should say something like "**Stable**: the component is ready for general availability."  The form implies no stage before "Stable" should ever be used in production, but there are times Beta could be used.  The later implies this component has had a "production" release, thus making it safe to use.

Author: Makes sense to me

Author: PR updated.

## Gated features

> Some features are hidden behind feature gates before they are ready for general availability. In such cases, even if a component is already stable, the features behind a feature gate should be considered **Experimental**.

Reviewer 3: I think the correct answer is, look for the lifecycle stage? We should not consider any feature behind a gate as experimental.

Also I like that our stages map with https://github.com/open-telemetry/opentelemetry-collector/blob/v0.49.0/service/featuregate/README.md#feature-lifecycle where GA == stable.

> Some features are hidden behind feature gates before they are ready for general availability. In such cases, even if a component is already stable, the features behind a feature gate should be considered **Experimental**.

Reviewer: If kept, Experimental should be updated to either Alpha or Beta

Reviewer 5: So if there's e.g. a stable component that has a feature gate, the said feature should be considered "**Alpha**" rather than "**Experimental**"?

Reviewer: I think the whole line was actually removed based on @bogdandrutu's feedback and this comment somehow didn't marked "Outdated"
